i6_models/parts/factored_hybrid/diphone.py

```diff
@@ -0,0 +1,167 @@
+__all__ = [
+    "DiphoneLogitsV1",
+    "DiphoneProbsV1",
+    "DiphoneBackendV1Config",
+    "DiphoneBackendV1",
+]
+
+from dataclasses import dataclass
+from typing import Callable, Optional, Union
+
+import torch
+from torch import nn, Tensor
+import torch.nn.functional as F
+
+from i6_models.config import ModelConfiguration
+
+from .util import PhonemeStateClassV1, get_center_dim, get_mlp
+
+
+@dataclass
+class DiphoneLogitsV1:
+    """outputs of a diphone factored hybrid model"""
+
+    embeddings_left_context: Tensor
+    """the embedded left context values"""
+
+    output_center: Tensor
+    """center output"""
+
+    output_left: Tensor
+    """left output"""
+
+
+class DiphoneProbsV1(DiphoneLogitsV1):
+    """marker class indicating that the output tensors store probabilities and not logits"""
+
+
+@dataclass
+class DiphoneBackendV1Config(ModelConfiguration):
+    """
+    Attributes:
+        context_mix_mlp_dim: inner dimension of the context mixing MLP layers
+        context_mix_mlp_num_layers: how many hidden layers on the MLPs there should be
+        dropout: dropout probabilty
+        left_context_embedding_dim: embedding dimension of the left context
+            values. Good choice is in the order of n_contexts.
+        n_contexts: the number of raw phonemes/acoustic contexts
+        num_hmm_states_per_phone: the number of HMM states per phoneme
+        num_inputs: input dimension of the backend, must match w/ output dimension
+            of main encoder (e.g. Conformer)
+        phoneme_state_class: the phoneme state augmentation to apply
+        activation: activation function to use in the context mixing MLP.
+    """
+
+    left_context_embedding_dim: int
+    n_contexts: int
+    num_hmm_states_per_phone: int
+    phoneme_state_class: Union[int, PhonemeStateClassV1]
+
+    activation: Callable[[], nn.Module]
```

Comment:
would that be a function that takes no input but returns a Module?

In other places I have seen the typing
```suggestion
    activation: Union[nn.Module, Callable[[torch.Tensor], torch.Tensor]]
```
which seems more appropriate to me

Reply:
offline discussion: With the fear of a statefull activation module this is indeed intended as a function that creates new instances every time it is called.